### PR TITLE
Enable RANSAC in triangulation

### PR DESCRIPTION
**Pull request by travisdriver**
Enabled RANSAC in the triangulation module.

### CODE REVIEW

**Comment by travisdriver** [comment]
The results on Skydio-32 look promising (will post other results once the benchmarking finishes):

# Before

Metric name | Value
-- | --
number_cameras | 32
number_tracks_unfiltered | 5724
mean_3d_track_lengths_unfiltered | 2.67313
mean_reprojection_errors_unfiltered_px | 10.9595
mean_rotation_error_angle_deg | 7.326
mean_translation_error_distance | 0.651919
mean_translation_angle_error_deg | 11.4042
number_tracks_filtered | 1256
mean_3d_track_lengths_filtered | 2.08758
mean_reprojection_errors_filtered_px | 1.39266

# After

Metric name | Value
-- | --
number_cameras | 32
number_tracks_unfiltered | 3482
mean_3d_track_lengths_unfiltered | 2.04451
mean_reprojection_errors_unfiltered_px | 0.669101
mean_rotation_error_angle_deg | 6.7561
mean_translation_error_distance | 0.920434
mean_translation_angle_error_deg | 12.0292
number_tracks_filtered | 3445
mean_3d_track_lengths_filtered | 2.04325
mean_reprojection_errors_filtered_px | 0.641061

**Comment by johnwlambert** [comment]
Thanks for the PR, Travis. Cool to see. 

I'm curious about why the average camera distance from GT goes from 0.65 to 0.9. Could you try sharing what the visualizations look like before and after the change?

**Comment by johnwlambert** [comment]
Travis, is there any way you could share some examples of how RANSAC is doing? Like for some visualizations of tracks (using our DA track viz script that dumps vertically stacked patches), could we see which ones RANSAC is labeling as outliers? Do the outliers look correct in practice? I think having GT metrics here on 3d GT will help too : - )

**Comment by travisdriver** [comment]
> Travis, is there any way you could share some examples of how RANSAC is doing? Like for some visualizations of tracks (using our DA track viz script that dumps vertically stacked patches), could we see which ones RANSAC is labeling as outliers? Do the outliers look correct in practice? I think having GT metrics here on 3d GT will help too : - )

Which GT metrics are you referring to?

**Comment by johnwlambert** [comment]
> > Travis, is there any way you could share some examples of how RANSAC is doing? Like for some visualizations of tracks (using our DA track viz script that dumps vertically stacked patches), could we see which ones RANSAC is labeling as outliers? Do the outliers look correct in practice? I think having GT metrics here on 3d GT will help too : - )
> 
> Which GT metrics are you referring to?

Oh I meant the track precision/recall from the 3d model -- I think we might need a bit more DA metrics before merging this DA PR

**Comment by travisdriver** [comment]
@johnwlambert Once you land #385, feel free to close this and open a fresh branch with the same changes.

**Comment by johnwlambert** [comment]
How does the dashboard look for this PR Travis? Could you share the file if you have created it?

**Comment by travisdriver** [comment]
Here are the results for 100 RANSAC iterations with a 5 pixel threshold.

I think door-12 may be so "converged" that small perturbations in the results make it look like it's doing really bad in the dashboard. For example, the ~1000% increase in translation error is a result of it going from 0.002 to 0.02. Nevertheless, the results are still not great.

I still think there's something peculiar going on. I would have at least expected the number of unfiltered tracks to increase (since were no longer rejecting entire tracks), but this is not the case. Also, the mean track length seems to have decreased substantially in most cases, suggesting that RANSAC is rejecting a significant amount of track correspondences (maybe we need to increase the pixel threshold).

I will investigate and debug. 

image

**Comment by travisdriver** [comment]
Actually I guess I would expect the number of unfiltered tracks to decrease since the default reprojection error threshold is 100 pixels on `master`...

**Comment by travisdriver** [comment]
Here are the results with a RANSAC threshold of 100 pixels. 

My hypothesis is that the poses from the averaging module are not accurate enough to perform RANSAC at more reasonable thresholds (this is supported by the huge performance increase afforded by #225).

newplot (1)

**Comment by johnwlambert** [comment]
Hi @travisdriver, could you post the dashboard to show the improvement on AstroNet with these changes?

**Comment by travisdriver** [comment]
newplot (7)

The decrease in performance for **Skydio-32** is likely just noise, as it has already converged, i.e., the decrease in the metrics is on the order of .1 and .01.

**Skydio-8** performed poorly on the CI but looked perfect on my machine; this is likely a symptom of #323. Here is the final reconstruction from running Skydio-8 locally:

Screenshot from 2021-12-15 00-02-29

### Skydio-8 metrics **w/ RANSAC** from local run
Screenshot from 2021-12-15 09-58-14

### Skydio-8 metrics **w/o RANSAC** from local run
Screenshot from 2021-12-15 10-01-44

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:69]: Hi @travisdriver, thanks for the PR. Could you add docstrings explaining each of these params, e.g. `dyn_num_hypotheses_multiplier`?

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:89]: commented out code -- can it be removed?

File: gtsfm/data_association/data_assoc.py
Comment (johnwlambert) [inline gtsfm/data_association/data_assoc.py:98]: commented-out code -- can it be removed?

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (johnwlambert) [inline gtsfm/configs/deep_front_end_astronet.yaml:55]: could we add here what a value of `1` corresponds to, instead of what `0` corresponds to?

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (johnwlambert) [inline gtsfm/configs/deep_front_end_astronet.yaml:57]: @travisdriver does the RANSAC triangulation setting hurt the other benchmarks (besides AstroNet)? If so, do we know why?

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:80]: @travisdriver could we add a unit test on this?

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:67]: I'm more accustomed to "inlier_ratio" rather than "min_inlier_ratio" when computing the number of RANSAC hypotheses -- what does `min_inlier_ratio` mean here?

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:123]: just curious -- why do we not need this check anymore?

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:131]: Would be helpful to make this log message more descriptive, e.g. 
```python
logger.debug("Num RANSAC hypotheses for triangulation: %d", num_hypotheses)
```

File: gtsfm/configs/deep_front_end.yaml
Comment (johnwlambert) [inline gtsfm/configs/deep_front_end.yaml:55]: same here -- would be nice to explain what 1 corresponds to. If we are using RANSAC in this config, what do we need the new `deep_front_end_astronet` config for?

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (travisdriver) [reply gtsfm/configs/deep_front_end_astronet.yaml:57]: The RANSAC triangulation doesn't seem to be hurting any of the relevant benchmarks: see [this comment](https://github.com/borglab/gtsfm/pull/292#issuecomment-999275501). 

It significantly improves AstroNet with a more appropriate threshold of 100 pixels (since the averaged poses are still not very good). It's all a function of the accuracy of the averaged poses.

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:131]: This was just for debugging purposes; I will remove it.

File: gtsfm/configs/deep_front_end.yaml
Comment (travisdriver) [reply gtsfm/configs/deep_front_end.yaml:55]: AstroNet performs better with a higher RANSAC threshold (100 pixels as opposed to 10) due to the quality of the averaged poses.

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:123]: Because the checks are performed by `__check_ransac_params()` when `num_ransac_hypotheses()` is called.

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:67]: Yes, the name is actually misleading (I was following COLMAP's implementation). It's actually the assumed outlier ratio, i.e., a higher value will result in more iterations in order to achieve the defined confidence.

I will change the name and add a description to the docstring.

Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:67]: Thanks

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (johnwlambert) [inline gtsfm/configs/deep_front_end_astronet.yaml:1]: maybe we could add a comment to the top of this yaml file, explaining that the only change w.r.t. `deep_front_end.yaml` is the RANSAC triangulation reproj threshold (10x looser)

Comment (travisdriver) [reply gtsfm/configs/deep_front_end_astronet.yaml:3]: added

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (travisdriver) [reply gtsfm/configs/deep_front_end_astronet.yaml]: added

File: gtsfm/data_association/data_assoc.py
Comment (travisdriver) [reply gtsfm/data_association/data_assoc.py]: removed

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:82]: added

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:94]: added

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py]: removed

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py]: removed

**Comment by johnwlambert** [comment]
Thanks for the PR, Travis. I'm taking a look at the report now:

[visual_comparison_dashboard 3.html.zip](https://github.com/borglab/gtsfm/files/7810261/visual_comparison_dashboard.3.html.zip)

Just trying to figure out which changes are sort of random changes vs. deterministic ones. I think you have a better sense from me from having run the CI a few times on this branch.

Maybe if we fix the determinism bug first it will be a bit clearer. What are your thoughts?

**Comment by johnwlambert** [comment]
Hi @travisdriver the PR looks good. Do you have a sense for the new runtime? I'm curious if it's 1% slower, 10% slower, or 100%+ slower with the new RANSAC loop.

**Comment by travisdriver** [comment]
> Hi @travisdriver the PR looks good. Do you have a sense for the new runtime? I'm curious if it's 1% slower, 10% slower, or 100%+ slower with the new RANSAC loop.

I'm not sure what you're referring to; there's no increase in runtime according to the CI.

**Comment by johnwlambert** [comment]
> > Hi @travisdriver the PR looks good. Do you have a sense for the new runtime? I'm curious if it's 1% slower, 10% slower, or 100%+ slower with the new RANSAC loop.
> 
> I'm not sure what you're referring to; there's no increase in runtime according to the CI.

I was thinking since we were adding a new inner Python for-loop for our 2-view geometry estimation, so that could slow things down considerably. But I guess since we limit to just 100 hypotheses the effect on runtime is neglible?

**Comment by travisdriver** [comment]
> > > Hi @travisdriver the PR looks good. Do you have a sense for the new runtime? I'm curious if it's 1% slower, 10% slower, or 100%+ slower with the new RANSAC loop.
> > 
> > 
> > I'm not sure what you're referring to; there's no increase in runtime according to the CI.
> 
> I was thinking since we were adding a new inner Python for-loop for our 2-view geometry estimation, so that could slow things down considerably. But I guess since we limit to just 100 hypotheses the effect on runtime is neglible?

Yes, triangulation is a relatively inexpensive operation.

**Comment by travisdriver** [comment]
@johnwlambert What else needs to be done in order to merge this PR?

**Comment by johnwlambert** [comment]
Hi Travis, thanks for your patience here. I'm a bit confused about the performance on skydio-8 and skydio-32. I triggered the CI one more time to see if those regressions were random. @akshay-krishnan @ayushbaid could you all take a look? I'm fine with landing the new config just for astronet right away, though. 
[visual_comparison_dashboard.html.zip](https://github.com/borglab/gtsfm/files/7893074/visual_comparison_dashboard.html.zip)

I saw that the runtime went from 58 min to 2 hours 20 min just for one benchmark (deep front end RC3). I'm running again to see if that's repeatable.
Screen Shot 2022-01-18 at 6 41 26 PM

**Comment by johnwlambert** [comment]
I checked the dashboard for the second run, and still seeing some confusing things. Hopefully if we land the repeatability stuff first, it will be more clear for us
Screen Shot 2022-01-19 at 9 34 39 AM

**Comment by travisdriver** [comment]
> I saw that the runtime went from 58 min to 2 hours 20 min just for one benchmark (deep front end RC3). I'm running again to see if that's repeatable.

This is from the most recent commit on master; the benchmark runtime fluctuates a lot for some reason, but there are plenty of instances other than this branch where is takes >2 hours.

Screenshot from 2022-01-19 09-34-40

**Comment by johnwlambert** [comment]
> > I saw that the runtime went from 58 min to 2 hours 20 min just for one benchmark (deep front end RC3). I'm running again to see if that's repeatable.
> 
> This is from the most recent commit on master; the benchmark runtime fluctuates a lot for some reason, but there are plenty of instances other than this branch where is takes >2 hours.
> 
> Screenshot from 2022-01-19 09-34-40

I see, cool. Didn't realize runtime was fluctuating that much too. I wonder if the hardware specs are different between runs on the CI

**Comment by travisdriver** [comment]
> I checked the dashboard for the second run, and still seeing some confusing things. Hopefully if we land the repeatability stuff first, it will be more clear for us 

I don't think an increase in angular error from .3 to .8 should completely hold the PR, especially when there are significant improvements across the board. The repeatability issues effect all PRs, but it seems to only be holding up this one...



> I checked the dashboard for the second run, and still seeing some confusing things. Hopefully if we land the repeatability stuff first, it will be more clear for us Screen Shot 2022-01-19 at 9 34 39 AM

Thanks for looking at the benchmarks. What's confusing from the output? A .5 degree increase in the mean rotation angle error seems insignificant compared to the other improvements.

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:100]: Hi Travis, could you add a reference for this dynamic multiplier? If COLMAP is using it, do you mind adding an inline comment w/ URL to the COLMAP source where they do it?

File: gtsfm/data_association/point3d_initializer.py
Comment (akshay-krishnan) [inline gtsfm/data_association/point3d_initializer.py:77]: I feel like TriangulationParam is not a good name for this enum. It means the same thing as `TriangulationOptions` but the enum is about what type of ransac to run. Maybe RansacMode or something is better.

File: gtsfm/data_association/point3d_initializer.py
Comment (akshay-krishnan) [inline gtsfm/data_association/point3d_initializer.py:80]: isnt that a really small ratio? do we have a min_num_inliers too? just to be sure this does not select tracks with one or two inliers?

File: gtsfm/data_association/point3d_initializer.py
Comment (akshay-krishnan) [inline gtsfm/data_association/point3d_initializer.py:102]: `num_hypotheses = max(min(self.max_num_hypotheses, dyn_num_hypotheses), self.min_num_hypothesis)` ?

File: gtsfm/data_association/point3d_initializer.py
Comment (akshay-krishnan) [inline gtsfm/data_association/point3d_initializer.py:100]: this works, but just for super clear readability could you add an additional pair of parenthesis? (A / B) * C instead of A / B * C

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:77]: I'm ok with it currently. I see what you mean Akshay, but I think `RansacMode.NO_RANSAC` would be sort of weird.

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:73]: smal typo here: "max" and "maximum"

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:77]: Maybe `SamplingMode` is better?

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:94]: marking as resolved

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:78]: I like `SamplingMode`

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:100]: COLMAP uses it, but all it does it multiply the number of dynamic hypotheses by a constant so I'm indifferent about whether it warrants a reference

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:81]: Yes, a smaller ratio will result in a larger number of dynamic hypotheses.

Keep in mind that the measurements are the cameras (with the corresponding keypoint), so a track with one inlier would result in a underdetermined problem (which we check for and triangulation would fail anyways). Two inliers is allowed, but tracks are discarded according to `min_track_length`.

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py]: This is better thank you

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:78]: or `TriangulationSamplingMode`?

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py]: i updated it, marking as resolved

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py]: i updated it, marking as resolved

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py]: i updated it, marking as resolved

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (ayushbaid) [inline gtsfm/configs/deep_front_end_astronet.yaml:1]: Can we name it something which mentions this attribute value being high instead of tagging it with a database?

File: gtsfm/configs/scene_optimizer_unit_test_config.yaml
Comment (ayushbaid) [inline gtsfm/configs/scene_optimizer_unit_test_config.yaml:53]: Why change to ransac in the unit test?

**Review by akshay-krishnan — APPROVED** [review]

File: gtsfm/data_association/point3d_initializer.py
Comment (akshay-krishnan) [reply gtsfm/data_association/point3d_initializer.py:81]: Okay, I wonder if a higher min inlier ratio produces better hypothesis? It would definitely save some compute.

File: gtsfm/data_association/point3d_initializer.py
Comment (ayushbaid) [inline gtsfm/data_association/point3d_initializer.py:72]: Can you elaborate what this is in the PR comments?

Also good to have an example computation in the docstring for the number of hypothesis.

File: gtsfm/data_association/point3d_initializer.py
Comment (ayushbaid) [inline gtsfm/data_association/point3d_initializer.py:121]: can we rename to triangulation_options to prevent ambiguity?

**Review by ayushbaid — APPROVED** [review]
LGTM except few minor comments.

Do you want to wait for the repeatability issues in 1dsfm to land before using it in astronet, or do you want to start using it regardless?

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (johnwlambert) [reply gtsfm/configs/deep_front_end_astronet.yaml:1]: I agree, I think this would improve clarity

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:121]: I feel like maybe that would be too wordy/long. Just my two cents.

File: gtsfm/configs/scene_optimizer_unit_test_config.yaml
Comment (johnwlambert) [reply gtsfm/configs/scene_optimizer_unit_test_config.yaml:53]: I didn't quite follow here Ayush -- I believe we were using NO_RANSAC before, and the test is still using NO_RANSAC now.

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:83]: thanks, marking as resolved

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:72]: @travisdriver we could add the following math:
```python
The RANSAC module defaults to 2749 iterations, computed as 
np.log(1-0.9999) / np.log( 1 - 0.1 **2) * 3 = 2749.3
```

Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:78]: I added this to the docstring

File: gtsfm/data_association/point3d_initializer.py
Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py]: I updated this, marking as resolved

**Review by johnwlambert — APPROVED** [review]
LGTM

File: gtsfm/data_association/point3d_initializer.py
Comment (travisdriver) [reply gtsfm/data_association/point3d_initializer.py:131]: I think just `options` is sufficient since there aren't any similar attributes.

File: gtsfm/configs/deep_front_end_astronet.yaml
Comment (travisdriver) [reply gtsfm/configs/deep_front_end_astronet.yaml:1]: AstroNet is the only dataset that uses the config by default, so it seems fitting to name it accordingly.

**Pull request merged**
